napari/_qt/qt_viewer.py

```diff
@@ -58,6 +61,125 @@
 from ..utils.io import imsave_extensions
 
 
+def _npe2_decode_selected_filter(
+    ext_str: str, selected_filter: str, writers: List[Any]
+) -> Optional[str]:
+    """When npe2 can be imported, resolves a selected file extension
+    string into a specific writer. Otherwise, returns None.
+
+    Returns the writer command that should be invoked to save data.
+    """
+    # When npe2 is not present, `writers` is expected to be an empty list,
+    # `[]`. This function will return None.
+
+    for entry, writer in zip(
+        ext_str.split(";;"),
+        writers,
+    ):
+        if entry.startswith(selected_filter):
+            return writer.command
+    return None
+
+
+def _npe2_file_extensions_string_for_layers(
+    layers: List[Layer] | LayerList,
+) -> Tuple[Optional[str], List[Any]]:
+    """When npe2 can be imported, returns an extension string and the list
+    of corresponding writers. Otherwise returns (None,[]).
+
+    The extension string is a ";;" delimeted string of entries. Each entry
+    has a brief description of the file type and a list of extensions.
+
+    The writers, when provided, are the
+    npe2.manifest.io.WriterContribution objects. There is one writer per
+    entry in the extension string.
+    """
+    try:
+        import npe2
+    except ImportError:
+        return None, []
+
+    layer_types = [layer.as_layer_data_tuple()[2] for layer in layers]
```

Comment:
Maybe open and issue to :

  - [x] have as_layer_data_tuple() return a names tuple. https://github.com/napari/napari/issues/3492

Reply:
Set comprehension ? Or do we actually care about the cardinality of each type and their order ? 
```suggestion
    layer_types = {layer.as_layer_data_tuple()[2] for layer in layers}
```

Reply:
We do care about cardinality here.

Reply:
for some reason, I can't add a new suggestion here... but this should be

```suggestion
    layer_types = [layer._type_string for layer in layers]
```

(no need to get the full layer state tuple if you just need the string)

Reply:
^ Edited to make suggestion.